firebase-firestore/src/main/java/com/google/firebase/firestore/local/BundleSerializer.java

```diff
@@ -0,0 +1,350 @@
+
+package com.google.firebase.firestore.local;
+
+import android.util.Base64;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+
+import com.google.firebase.Timestamp;
+import com.google.firebase.firestore.core.Bound;
+import com.google.firebase.firestore.core.FieldFilter;
+import com.google.firebase.firestore.core.Filter;
+import com.google.firebase.firestore.core.OrderBy;
+import com.google.firebase.firestore.core.Query;
+import com.google.firebase.firestore.model.Document;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.FieldPath;
+import com.google.firebase.firestore.model.ObjectValue;
+import com.google.firebase.firestore.model.ResourcePath;
+import com.google.firebase.firestore.model.SnapshotVersion;
+import com.google.firebase.firestore.model.Values;
+import com.google.firebase.firestore.remote.RemoteSerializer;
+import com.google.firestore.v1.ArrayValue;
+import com.google.firestore.v1.MapValue;
+import com.google.firestore.v1.Value;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.NullValue;
+import com.google.type.LatLng;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/** A JSON serializer to deserialize Firestore Bundles. */
+/* package= */ class BundleSerializer {
+
+  private final RemoteSerializer remoteSerializer;
+
+  public BundleSerializer(RemoteSerializer remoteSerializer) {
+    this.remoteSerializer = remoteSerializer;
+  }
+
+  public NamedQuery fromNamedQuery(String json) throws JSONException {
+    JSONObject JSONObject = new JSONObject(json);
+    String name = JSONObject.getString("name");
+    Query query = decodeQuery(JSONObject.getJSONObject("bundledQuery"));
+    SnapshotVersion readTime = decodeSnapshotVersion(JSONObject.getJSONObject("readTime"));
+    return new NamedQuery(name, query, readTime);
+  }
+
+  public BundleMetadata fromBundleMetadata(String json) throws JSONException {
+    JSONObject jsonObject = new JSONObject(json);
+    String bundleId = jsonObject.getString("id");
+    int version = jsonObject.getInt("version");
+    SnapshotVersion createTime = decodeSnapshotVersion(jsonObject.getJSONObject("createTime"));
+    int totalDocuments = jsonObject.getInt("totalDocuments");
+    long totalBytes = jsonObject.getLong("totalBytes");
+    return new BundleMetadata(bundleId, version, createTime, totalDocuments, totalBytes);
+  }
+
+  public BundledDocumentMetadata fromBundledDocumentMetadata(String json) throws JSONException {
+    JSONObject jsonObject = new JSONObject(json);
+    DocumentKey key = DocumentKey.fromName(jsonObject.getString("name"));
+    SnapshotVersion readTime = decodeSnapshotVersion(jsonObject.getJSONObject("readTime"));
+    boolean exists = jsonObject.optBoolean("exists", false);
+    JSONArray queriesJson = jsonObject.optJSONArray("queries");
```

Comment:
An empty array is different from a missing array. Are you sure you want to conflate these two notions here? If so, it may be worth a comment on why that's OK.

Reply:
It seems safer to treat a missing array and an empty array the same, despite the fact that a missing array might point to a problem in Server SDKs. Added comment.